src/main/java/org/jenkinsci/plugins/configfiles/custom/security/CustomizedCredentialMapping.java

```diff
@@ -0,0 +1,107 @@
+ LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ THE SOFTWARE.
+ */
+package org.jenkinsci.plugins.configfiles.custom.security;
+
+import com.cloudbees.plugins.credentials.common.IdCredentials;
+import com.cloudbees.plugins.credentials.common.StandardListBoxModel;
+import com.cloudbees.plugins.credentials.common.StandardUsernameCredentials;
+import com.cloudbees.plugins.credentials.domains.DomainRequirement;
+import hudson.Extension;
+import hudson.model.AbstractDescribableImpl;
+import hudson.model.Descriptor;
+import hudson.model.Item;
+import hudson.model.ItemGroup;
+import hudson.model.Queue;
+import hudson.security.ACL;
+import hudson.security.AccessControlled;
+import hudson.util.ListBoxModel;
+import jenkins.model.Jenkins;
+import org.apache.commons.lang.StringUtils;
+import org.jenkinsci.plugins.plaincredentials.StringCredentials;
+import org.kohsuke.stapler.AncestorInPath;
+import org.kohsuke.stapler.DataBoundConstructor;
+import org.kohsuke.stapler.QueryParameter;
+
+import java.io.Serializable;
+import java.util.Collections;
+import java.util.List;
+
+public class CustomizedCredentialMapping extends AbstractDescribableImpl<CustomizedCredentialMapping> implements Serializable {
+
+    private final String tokenKey;
+    private final String credentialsId;
+
+    @DataBoundConstructor
+    public CustomizedCredentialMapping(String tokenKey, String credentialsId) {
+        this.tokenKey = tokenKey;
+        this.credentialsId = credentialsId;
+    }
+
+    public String getTokenKey() {
+        return tokenKey;
+    }
+
+    public String getCredentialsId() {
+        return credentialsId;
+    }
+
+    public Descriptor<CustomizedCredentialMapping> getDescriptor() {
+        return DESCRIPTOR;
+    }
+
+    private static final DescriptorImpl DESCRIPTOR = new DescriptorImpl();
+
+    @Extension
+    public static class DescriptorImpl extends Descriptor<CustomizedCredentialMapping> {
+
+        public ListBoxModel doFillCredentialsIdItems(@AncestorInPath ItemGroup context, @QueryParameter String tokenKey) {
+            AccessControlled _context = (context instanceof AccessControlled ? (AccessControlled) context : Jenkins.get());
+            if (_context == null || !_context.hasPermission(Item.CONFIGURE)) {
+                return new StandardListBoxModel().includeCurrentValue(tokenKey);
+            }
+
+            List<DomainRequirement> domainRequirements = Collections.emptyList();
+            if (StringUtils.isNotBlank(tokenKey)) {
+                domainRequirements = Collections.singletonList(new TokenKeyRequirement(tokenKey));
+            }
+
+            StandardListBoxModel standardListBoxModel = new StandardListBoxModel();
+            standardListBoxModel.includeAs(
+                    context instanceof Queue.Task ? ((Queue.Task) context).getDefaultAuthentication() : ACL.SYSTEM,
```

Comment:
@daniel-beck is this and line 97 what you're concerned with? If so. I basically lifted this code from .../maven/security/ServerCredentialMapping.java and properties/security/PropertiesCredentialMapping.java